### ISSUE TITLE
[v6r20] Problems with BulkSubmission in Transformations

### ISSUE THREAD

**Issue by andresailer** [DIRACGrid/DIRAC#3707]
Besides the problem in #3706 

I tried enabling BulkSubmission in The WorkflowTaskAgent, but it seems that right now this is not fully compatible with my existing transformations.
In the Workflow XML of the transformation some parameters are defined with Dummy Values, e.g.:
JOB_ID with 00012345 and also JobName with "Name"
```
<Parameter name="JOB_ID" type="string" linked_module="" linked_parameter="" in="True" out="False" description="ProductionJobID"><value><![CDATA[00012345]]></value></Parameter>
<Parameter name="JobName" type="JDL" linked_module="" linked_parameter="" in="True" out="False" description="Name of Job"><value><![CDATA[Name]]></value></Parameter>
```

When the BulkSubmission happens these parameters are not overwritten as is the case for single submission, but instead appended, so in the JDL these are now lists
```
  Parameter.JobName =
  {
    "Name",
    00401094
  };
  JobName =
  {
    "Name",
    00401094
  };
  Parameter.JOB_ID =
  {
    00012345,
    00000040
  };
```
Which causes problems later on, when I use the Job ID, which is no longer an integer as before.
Also the jobname is:
`{ Name, 00401094 }`
instead of something like:
`00401094_00000033`

**Comment by atsareg** [comment]
Looking at it

**Comment by atsareg** [comment]
This is a "normal" behavior. When a job is defined with a JDL parameter and also with a sequence of the same name, then the two are merged together. This is quite handy for example for the InputSandbox where files to be uploaded are defined ones for all the jobs in the InputSandbox parameter and then a variable list of LFN's can be added as an InputSandbox sequence of parameters. As a result each job has a list of InputSandbox files containing common and individual files. This is a general Job API behavior, not specific to the Transformation.

However, we need JobName and JOB_ID placeholders because later on the values for these parameters will be added with the "setValue" semantics which assumes that the placeholders are already in place. But to avoid conflicts like you describe these parameters should not be of the JDL type which makes them added to the job JDL. 

In the current implementation of the Job.py API there is a bug that workflow non-JDL parameters and sequence parameters are merged together always without verification if the workflow parameter is of the JDL type. This will be corrected. Sequence parameters are always JDL. So, in your case just the JobName workflow parameter should be made of the "string" type. 

But you are speaking about the already existing transformations. Changing the transformation body is certainly not easy. May be, we should just drop this merging semantics and make always parameters from sequences override the JDL parameters of the same name. Or only merge if the JDL parameter has a list-type value. Comments ?

**Comment by andresailer** [comment]
Another problem that shows up is that the "ProductionOutputData" and some other parameters that come via the OutputDataPolicy are not present in the workflow_commons dictionary.
Probably because these parameters are not added to the JDL like it is done in `__prepareTasks`
https://github.com/DIRACGrid/DIRAC/blob/5e4ab28334e7359fdb25c0962ad6f021acccdd38/TransformationSystem/Client/TaskManager.py#L650-L661

**Comment by atsareg** [comment]
Indeed, these parameters were added to the JDL but not to the workflow. Now this should be done

**Comment by andresailer** [comment]
Solved with #3709
